File: desktop/crates/catlog-gui/src/sounds.rs
//! The sounds: every moment the desk makes one at has its own choice,
//! kept on this machine: none, one of the shipped cat sounds, or a file
//! of the keeper's own. The shipped ones, each moment its own cat: a
//! short meow for a tick, a purr for a day of chores done, a chorus of
//! meows for a ladder climbed, a meow over a purr for an adoption. The
//! recordings are CC0 and public domain from Wikimedia Commons, see
//! `assets/sounds/LICENSES.md`; they ship with the app.

use std::path::{Path, PathBuf};

use catlog_core::Catalog;

use crate::l10n::L10n;

/// Plays a sound.
pub trait Sounder {
    fn play(&mut self, sound: Vec<u8>);
}

pub static TICK: &[u8] = include_bytes!("../../../../assets/sounds/tick.wav");
pub static PURR: &[u8] = include_bytes!("../../../../assets/sounds/purr.wav");
pub static CHORUS: &[u8] = include_bytes!("../../../../assets/sounds/chorus.wav");
pub static PARTY: &[u8] = include_bytes!("../../../../assets/sounds/party.wav");

/// The moments a sound belongs to.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Cheer {
    Tick,
    DayDone,
    Ladder,
    Adoption,
}

impl Cheer {
    pub const ALL: [Cheer; 4] = [Cheer::Tick, Cheer::DayDone, Cheer::Ladder, Cheer::Adoption];

    /// The local setting the choice is kept under; the phone's key.
    pub fn key(self) -> &'static str {
        match self {
            Cheer::Tick => "sound:tick",
            Cheer::DayDone => "sound:dayDone",
            Cheer::Ladder => "sound:ladder",
            Cheer::Adoption => "sound:adoption",
        }
    }

    /// The moment's name on the Settings page.
    pub fn label(self, t: &L10n) -> &'static str {
        match self {
            Cheer::Tick => t.chore_tick_label(),
            Cheer::DayDone => t.all_done_today(),
            Cheer::Ladder => t.achievements_title(),
            Cheer::Adoption => t.toast_kind_adoptions(),
        }
    }

    /// The sound a moment ships with.
    pub fn default_preset(self) -> Preset {
        match self {
            Cheer::Tick => Preset::Meow,
            Cheer::DayDone => Preset::Purr,
            Cheer::Ladder => Preset::Chorus,
            Cheer::Adoption => Preset::Party,
        }
    }
}

/// The shipped sounds.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Preset {
    Meow,
    Purr,
    Chorus,
    Party,
}

impl Preset {
    pub const ALL: [Preset; 4] = [Preset::Meow, Preset::Purr, Preset::Chorus, Preset::Party];

    /// The value kept in the setting; the phone's spelling.
    pub fn name(self) -> &'static str {
        match self {
            Preset::Meow => "meow",
            Preset::Purr => "purr",
            Preset::Chorus => "chorus",
            Preset::Party => "party",
        }
    }

    fn from_name(name: &str) -> Option<Preset> {
        Preset::ALL.into_iter().find(|p| p.name() == name)
    }

    pub fn label(self, t: &L10n) -> &'static str {
        match self {
            Preset::Meow => t.sound_meow(),
            Preset::Purr => t.sound_purr(),
            Preset::Chorus => t.sound_chorus(),
            Preset::Party => t.sound_party(),
        }
    }

    pub fn bytes(self) -> &'static [u8] {
        match self {
            Preset::Meow => TICK,
            Preset::Purr => PURR,
            Preset::Chorus => CHORUS,
            Preset::Party => PARTY,
        }
    }
}

/// The sound a moment ships with.
pub fn cheer_sound(cheer: Cheer) -> &'static [u8] {
    cheer.default_preset().bytes()
}

/// What plays at a moment.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SoundChoice {
    None,
    Preset(Preset),
    Own(PathBuf),
}

impl SoundChoice {
    /// The choice's name: None, the preset, or the own file's name.
    pub fn label(&self, t: &L10n) -> String {
        match self {
            SoundChoice::None => t.alert_none().to_string(),
            SoundChoice::Preset(p) => p.label(t).to_string(),
            SoundChoice::Own(path) => path
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_default(),
        }
    }

    /// The bytes to play, read now for an own file; none for none or a
    /// file that is gone.
    pub fn bytes(&self) -> Option<Vec<u8>> {
        match self {
            SoundChoice::None => None,
            SoundChoice::Preset(p) => Some(p.bytes().to_vec()),
            SoundChoice::Own(path) => std::fs::read(path).ok(),
        }
    }
}

/// The choice for a moment. Before the choices existed one switch
/// silenced every cheer; a machine that had it off stays silent.
pub fn sound_for(store: &Catalog, cheer: Cheer) -> SoundChoice {
    let Some(raw) = store.local_setting(cheer.key()) else {
        let legacy_off = ["celebrationSound", "celebrations"]
            .iter()
            .any(|k| store.local_setting(k).as_deref() == Some("off"));
        return if legacy_off {
            SoundChoice::None
        } else {
            SoundChoice::Preset(cheer.default_preset())
        };
    };
    if raw == "none" {
        return SoundChoice::None;
    }
    if let Some(path) = raw.strip_prefix("file:") {
        return SoundChoice::Own(PathBuf::from(path));
    }
    SoundChoice::Preset(Preset::from_name(&raw).unwrap_or(cheer.default_preset()))
}

pub fn set_sound(store: &Catalog, cheer: Cheer, choice: &SoundChoice) {
    let raw = match choice {
        SoundChoice::None => "none".to_string(),
        SoundChoice::Preset(p) => p.name().to_string(),
        SoundChoice::Own(path) => format!("file:{}", path.display()),
    };
    let _ = store.set_local_setting(cheer.key(), &raw);
}

/// The file kinds the player can read.
pub const SOUND_FILES: &[&str] = &["wav", "mp3", "flac", "ogg"];

/// Keeps a copy of a picked file under `dir`, so the choice outlives
/// the file it came from.
pub fn keep_own(dir: &Path, cheer: Cheer, source: &Path) -> std::io::Result<PathBuf> {
    let sounds = dir.join("sounds");
    std::fs::create_dir_all(&sounds)?;
    let name = match source.extension() {
        Some(ext) => format!(
            "{}.{}",
            cheer.key().trim_start_matches("sound:"),
            ext.to_string_lossy()
        ),
        None => cheer.key().trim_start_matches("sound:").to_string(),
    };
    let target = sounds.join(name);
    std::fs::copy(source, &target)?;
    Ok(target)
}

/// The speakers, through rodio, on a thread of their own.
pub struct Speakers;

impl Sounder for Speakers {
    fn play(&mut self, sound: Vec<u8>) {
        std::thread::spawn(move || {
            let Ok(sink) = rodio::DeviceSinkBuilder::open_default_sink() else {
                return;
            };
            if let Ok(player) = rodio::play(sink.mixer(), std::io::Cursor::new(sound)) {
                player.sleep_until_end();
            }
        });
    }
}

/// Keeps what was played, by length.
#[derive(Debug, Default)]
pub struct RecordingSounder {
    pub played: std::sync::Arc<std::sync::Mutex<Vec<usize>>>,
}

impl Sounder for RecordingSounder {
    fn play(&mut self, sound: Vec<u8>) {
        if let Ok(mut played) = self.played.lock() {
            played.push(sound.len());
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_moment_has_its_own_short_wave() {
        for cheer in Cheer::ALL {
            let bytes = cheer_sound(cheer);
            assert!(bytes.starts_with(b"RIFF"), "{cheer:?}");
            assert!(
                bytes.len() > 10_000 && bytes.len() < 400_000,
                "{cheer:?}: short"
            );
        }
        assert!(cheer_sound(Cheer::Tick).len() < cheer_sound(Cheer::DayDone).len());
    }

    #[test]
    fn a_moment_keeps_its_choice_and_the_old_switch_is_honoured() {
        let dir = tempfile::tempdir().unwrap();
        let store = Catalog::open(dir.path()).unwrap();
        assert_eq!(
            sound_for(&store, Cheer::Tick),
            SoundChoice::Preset(Preset::Meow)
        );
        set_sound(&store, Cheer::Tick, &SoundChoice::None);
        assert_eq!(sound_for(&store, Cheer::Tick), SoundChoice::None);
        assert!(sound_for(&store, Cheer::Tick).bytes().is_none());
        set_sound(&store, Cheer::DayDone, &SoundChoice::Preset(Preset::Party));
        assert_eq!(
            sound_for(&store, Cheer::DayDone).bytes().unwrap().len(),
            PARTY.len()
        );
        // An own file is copied beside the data and read from there.
        let source = dir.path().join("mine.wav");
        std::fs::write(&source, TICK).unwrap();
        let kept = keep_own(dir.path().join("data").as_path(), Cheer::Ladder, &source).unwrap();
        assert_eq!(kept, dir.path().join("data/sounds/ladder.wav"));
        set_sound(&store, Cheer::Ladder, &SoundChoice::Own(kept.clone()));
        assert_eq!(sound_for(&store, Cheer::Ladder), SoundChoice::Own(kept));
        assert_eq!(
            sound_for(&store, Cheer::Ladder).bytes().unwrap().len(),
            TICK.len()
        );
        // A value the app does not know falls back to the moment's own.
        let _ = store.set_local_setting(Cheer::Adoption.key(), "trumpet");
        assert_eq!(
            sound_for(&store, Cheer::Adoption),
            SoundChoice::Preset(Preset::Party)
        );
        // The switch from before the choices: off keeps every unset moment silent.
        let old = Catalog::open(&dir.path().join("old")).unwrap();
        let _ = old.set_local_setting("celebrationSound", "off");
        for cheer in Cheer::ALL {
            assert_eq!(sound_for(&old, cheer), SoundChoice::None, "{cheer:?}");
        }
        set_sound(&old, Cheer::Tick, &SoundChoice::Preset(Preset::Purr));
        assert_eq!(
            sound_for(&old, Cheer::Tick),
            SoundChoice::Preset(Preset::Purr)
        );
        assert_eq!(sound_for(&old, Cheer::Ladder), SoundChoice::None);
    }
}

File: lib/src/sounds.dart
import 'dart:async';
import 'dart:io';

import 'package:audioplayers/audioplayers.dart';
import 'package:catalog_core/catalog_core.dart';
import 'package:file_picker/file_picker.dart';
import 'package:path_provider/path_provider.dart';

import 'celebration.dart' show Cheer, firstOrDone;
import '../l10n/app_localizations.dart';

/// Every moment the app makes a sound at has its own choice, kept on
/// this device: none, one of the shipped cat sounds, or a file of the
/// keeper's own. A choice is made on the Settings page and heard there
/// as it is picked.
sealed class SoundChoice {
  const SoundChoice();
}

class NoSound extends SoundChoice {
  const NoSound();
}

class PresetSound extends SoundChoice {
  final Preset preset;
  const PresetSound(this.preset);
}

class OwnSound extends SoundChoice {
  final String path;
  const OwnSound(this.path);
}

/// The shipped sounds, see assets/sounds/LICENSES.md: a purr and a
/// party meow from Wikimedia Commons, three calls of one cat called
/// Socke, and a chorus made of those three.
enum Preset { purr, chorus, party, meep, mrrp, mrrr }

String presetAsset(Preset preset) => switch (preset) {
  Preset.purr => 'sounds/purr.wav',
  Preset.chorus => 'sounds/chorus.wav',
  Preset.party => 'sounds/party.wav',
  Preset.meep => 'sounds/socke1.wav',
  Preset.mrrp => 'sounds/socke2.wav',
  Preset.mrrr => 'sounds/socke3.wav',
};

/// The sound a moment has until the keeper picks another.
Preset defaultPreset(Cheer moment) => switch (moment) {
  Cheer.tick => Preset.meep,
  Cheer.dayDone => Preset.purr,
  Cheer.ladder => Preset.chorus,
  Cheer.adoption => Preset.party,
};

String soundKey(Cheer moment) => 'sound:${moment.name}';

/// The choice for a moment. Before the choices existed one switch
/// silenced every cheer; a device that had it off stays silent.
SoundChoice soundFor(CatalogStore store, Cheer moment) {
  final raw = store.localSetting(soundKey(moment));
  if (raw == null) {
    final legacyOff =
        store.localSetting('celebrationSound') == 'off' ||
        store.localSetting('celebrations') == 'off';
    return legacyOff ? const NoSound() : PresetSound(defaultPreset(moment));
  }
  if (raw == 'none') return const NoSound();
  if (raw.startsWith('file:')) return OwnSound(raw.substring(5));
  final preset = Preset.values.where((p) => p.name == raw).firstOrNull;
  return PresetSound(preset ?? defaultPreset(moment));
}

void setSound(CatalogStore store, Cheer moment, SoundChoice choice) {
  final raw = switch (choice) {
    NoSound() => 'none',
    PresetSound(:final preset) => preset.name,
    OwnSound(:final path) => 'file:$path',
  };
  store.setLocalSetting(soundKey(moment), raw);
}

/// Plays a choice; nothing for none, and nothing when the device has
/// no audio to give.
Future<void> playSound(SoundChoice choice) async {
  final source = switch (choice) {
    NoSound() => null,
    PresetSound(:final preset) => AssetSource(presetAsset(preset)),
    OwnSound(:final path) => DeviceFileSource(path),
  };
  if (source == null) return;
  final player = AudioPlayer();
  try {
    await player.setAudioContext(
      AudioContext(
        iOS: AudioContextIOS(category: AVAudioSessionCategory.ambient),
        android: const AudioContextAndroid(
          usageType: AndroidUsageType.game,
          audioFocus: AndroidAudioFocus.none,
        ),
      ),
    );
    await player.play(source);
    // Released when the sound ends — or after a few seconds if the
    // platform never says so, or at once when the platform closes the
    // stream without an event.
    firstOrDone(
      player.onPlayerComplete,
      const Duration(seconds: 10),
    ).whenComplete(player.dispose);
  } catch (_) {
    // No audio device or platform quirk — the moment passes in silence.
    player.dispose();
  }
}

/// The moment's sound, as chosen.
Future<void> playMoment(CatalogStore store, Cheer moment) =>
    playSound(soundFor(store, moment));

/// Asks for an audio file and keeps a copy of it with the app, so the
/// choice outlives the file it came from. Null when nothing was picked.
Future<OwnSound?> pickOwnSound(Cheer moment) async {
  final picked = await FilePicker.platform.pickFiles(type: FileType.audio);
  final source = picked?.files.singleOrNull?.path;
  if (source == null) return null;
  final dir = await getApplicationSupportDirectory();
  final sounds = Directory('${dir.path}/sounds');
  await sounds.create(recursive: true);
  final dot = source.lastIndexOf('.');
  final extension = dot < 0 ? '' : source.substring(dot);
  final target = '${sounds.path}/${moment.name}$extension';
  await File(source).copy(target);
  return OwnSound(target);
}

/// The moment's name on the Settings page.
String momentLabel(AppLocalizations t, Cheer moment) => switch (moment) {
  Cheer.tick => t.choreTickLabel,
  Cheer.dayDone => t.allDoneToday,
  Cheer.ladder => t.achievementsTitle,
  Cheer.adoption => t.toastKindAdoptions,
};

String presetLabel(AppLocalizations t, Preset preset) => switch (preset) {
  Preset.purr => t.soundPurr,
  Preset.chorus => t.soundChorus,
  Preset.party => t.soundParty,
  Preset.meep => t.soundMeep,
  Preset.mrrp => t.soundMrrp,
  Preset.mrrr => t.soundMrrr,
};

/// The choice's name: None, the preset, or the own file's name.
String choiceLabel(AppLocalizations t, SoundChoice choice) => switch (choice) {
  NoSound() => t.alertNone,
  PresetSound(:final preset) => presetLabel(t, preset),
  OwnSound(:final path) => path.split(Platform.pathSeparator).last,
};
